elasticdl/python/data/parallel_transform.py

```diff
@@ -0,0 +1,123 @@
+import itertools
+from multiprocessing import Process, Queue
+
+
+def worker_loop(records, index_q, result_q, transform_fn):
+    """
+    The worker loop gets a index from the index queue, and
+    transform the records[index], then the result is put to
+    the result queue.
+
+    If a None index is got, the loop finishes.
+
+    :param records: a list object contains many records waiting to
+                    be transformed
+    :param index_q: a queue which generates the record index in records
+    :param result_q: a queue which stores the transform results
+    :param transform_fn: the transform function
+    """
+    while True:
+        index = index_q.get()
+        if index is None:
+            break
+        record = records[index]
+        res = transform_fn(record)
+        result_q.put((index, res))
+
+
+class ParallelTransform(object):
+    """
+    The ParallelTransform applies transform_fn to records with a number
+    of workers with low latency.
+
+    Worker process side:
+    1. get index from index queue
+    2. transform_fn(records[index])
+    3. put result to result queue
+
+    Main process side:
+    1. put index to index queue of workers in round-robin way
+    2. get transformed data from result queue
+
+    We want to achieve two things:
+    1. Data samples in input records are transformed with low latency.
+       We transform the data samples one by one. Once we call `next_data`
+       method to get a transformed result, the next index is put to a
+       index queue and triggers a worker process.
+
+    2. The output order is strictly the same with the input order.
+       We use a `cache_dict` to adjust the order, since the result order
+       getting from result queue is sequential sometimes.
+    """
+
+    def __init__(self, records, num_workers, transform_fn):
```

Comment:
`num_workers` is a little confusing in ElasticDL.
`num_parallel_processes` is better.

Reply:
Done